### PR TITLE
Streamline plugin model to better support external plugins

### LINKED ISSUE #35: [enhancement] Make plugin model more portable
## Description

With the current architecture, a user that authors a plugin must manually update the `/src/dts/Cargo.toml` file to include it. Additionally, all of the sample plugin crates are included there even though most of them aren't actually being built into the application. This negatively affects the portability of Freyja, which notably impacts the ability to include it in SDV blueprints. Ideally, plugin authors would not have to modify Freyja source code to include their external plugin.

Possible solutions to this problem:

- Utilize the build system to import external crates based on build config or environment variables. Possible tools or techniques would be usage of a build.rs file or something like [cargo-make](https://crates.io/crates/cargo-make) to define additional build tasks
- ~~Change to a dynamically loaded plugin model using tools such as [abi_stable](https://crates.io/crates/abi_stable/) or [wasmer](https://crates.io/crates/wasmer/)~~
- ~~Take advantage of the fact that all plugins are expected to communicate with external services and remove the plugin model entirely in favor of defining an API and requiring customers to integrate against that API instead. This might be difficult with the proxy interfaces though since these inherently need to know how to use certain communication protocols which might be customer-defined and not suitable for inclusion in the source code~~

## Acceptance criteria

- [x] Users can author plugins separately and must change only environment config to have Freyja build using their plugins

### PR DESCRIPTION
**Pull request by wilyle**
Streamlines the plugin model by auto-generating a crate at build time which refers to the plugins. This means that adding a new plugin no longer requires a code change to a Cargo.toml file, making it easier to use external plugins with Freyja.

Closes #35

### CODE REVIEW

File: docs/external-libs.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+Note that each trait implementation must live in a separate crate. You cannot put all of the implementations in one crate.
+
+## Freyja Dependency Generator
+
+The Freyja dependency generator helps with bringing in custom libraries and linking then with the main Freyja application. The dependency generator is an executable located in the `depgen` folder, and the generated files are placed in `depgen/__generated`. This folder contains some placeholder files so that cargo can check the workspace properly and so that the code to generate the files can be reduced. These placeholders will be part of the repo when you clone it, but changes will be ignored.
```
Comment (jorchiu) [inline docs/external-libs.md:13]: Please change "then" to "them" in "... and linking then with the main Freyja application".

File: docs/external-libs.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+**It is strongly recommended to use `cargo make` to interact with the project when possible** (e.g. `cargo make build`). This ensures that the dependency generator runs automatically and picks up your build configuration since all of the `cargo make` commands for this project will invoke the generator. The placeholder files that come with the repo are valid and the project can be built with plain cargo, but the default implementations are basic in-memory mocks which are generally only useful for prototyping. A few common cargo operations have equivalent `cargo make` tasks defined for them, such as `build`, `test`, and `fmt`. If you need to run the dependency generator manually you can run `cargo make freyja-depgen`. Note that the dependency generator relies on the build envinronment provided by `cargo make`, so it can't be run outside of the context. It will also require that all of the environment variables mentioned below are set.
```
Comment (jorchiu) [inline docs/external-libs.md:15]: Typo for environment  in "...relies on the build envinronment"

File: dts/Cargo.toml [hunk 3]
```diff
@@ -11,24 +11,14 @@ license = "MIT"
...
 async-trait = { workspace = true }
 time = { workspace = true }
 env_logger = { workspace = true }
 provider-proxy-selector = { path = "../provider_proxy_selector"}
-crossbeam = { workspace = true }
\ No newline at end of file
+crossbeam = { workspace = true }
+
+[dev-dependencies]
+in-memory-mock-cloud-adapter = { path = "../cloud_adapters/in_memory_mock_cloud_adapter" }
```
Comment (jorchiu) [inline dts/Cargo.toml:24]: Just curious why is this needed as a dev dependency if it's specified in the freyja-deps crate?

File: Makefile.toml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+[tasks.freyja-ci-build-and-test]
+dependencies = ["freyja-depgen", "cargo-build", "cargo-test"]
+
+[tasks.freyja-ci-code-analysis]
+dependencies = ["freyja-depgen", "cargo-check", "cargo-clippy", "cargo-fmt-check", "cargo-doctest", "cargo-doc"]
+
+# cargo commands
+[tasks.cargo-build]
+command = "cargo"
```
Comment (jorchiu) [inline Makefile.toml:30]: Does there need to be one for cargo audit?

File: Makefile.toml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+env_files = [
```
Comment (ashbeitz) [inline Makefile.toml:1]: Please add a copyright banner.  A comment describing the significance of this file would be good too.

File: depgen/__generated/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,3 @@
+pub use in_memory_mock_cloud_adapter::in_memory_mock_cloud_adapter::InMemoryMockCloudAdapter as CloudAdapterImpl;
```
Comment (ashbeitz) [inline depgen/__generated/src/lib.rs:1]: Please add a copyright banner.

File: depgen/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    let lib_file = Path::new(&workspace)
+        .join(GENERATED_DIR_RELATIVE_TO_WORKSPACE)
+        .join(LIBRS_FILE_PATH);
+
+    write_cargo_toml(&cargo_file)?;
+    write_lib(&lib_file)?;
+    Ok(())
+}
+
+fn write_cargo_toml<P>(path: &P) -> Result<(), String>
```
Comment (ashbeitz) [inline depgen/src/main.rs:32]: Please add a method comment.

Comment (ashbeitz) [reply depgen/src/main.rs:32]: For all methods.

File: depgen/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    let mut toml = fs::read_to_string(path)
+        .map_err(|e| format!("Unable to read file at {path:?}: {e:?}"))?
+        .parse::<Document>()
+        .map_err(|e| format!("Unable to parse Cargo.toml file: {e:?}"))?;
+
+    let mut dependencies = toml_edit::table();
+
+    // Digital Twin Adapter
+    let dt_adapter_package_name = get_env("FREYJA_DT_ADAPTER_PKG_NAME")?;
```
Comment (ashbeitz) [inline depgen/src/main.rs:46]: Please use constants for your env names.

**Review by ashbeitz — COMMENTED** [review]
Please take a look at my comments.

File: dts/Cargo.toml (hunk 3 above)
Comment (wilyle) [reply dts/Cargo.toml:24]: if you change the environment variables to use a different digital twin adapter then this is no longer referenced anywhere and the tests will not build. it's used for a test though, so I've put it in the dev-dependencies. Ideally I think it shouldn't be there, but that's not something to deal with in this PR

Comment (wilyle) [reply dts/Cargo.toml:24]: as a reminder, everything in `__generated` is a placeholder. I could make it empty instead, but I talked about this feature with Devin and he suggested that a working placeholder would be more intuitive. If this is causing confusion though, I can make the placeholder empty

File: docs/external-libs.md (hunk 2 above)
Comment (wilyle) [reply docs/external-libs.md:15]: odd that the spell checker didn't pick this up

File: Makefile.toml (hunk 4 above)
Comment (wilyle) [reply Makefile.toml:30]: technically no, it will work just fine as it is. not a bad idea to have one anyway

File: dts/Cargo.toml (hunk 3 above)
Comment (jorchiu) [reply dts/Cargo.toml:24]: I think it's good to have a working placeholder too, so you can keep the files in `__generated`

**Comment by wilyle** [comment]
This attempt at a solution turns out to have a fundamental issue with cargo-make since it runs `cargo metadata` before every command, which will fail if the auto-generated crate doesn't already exist. Thus the generation can't be done with cargo-make. Closed in favor of #39
